fix(tracing): direct OTel SDK setup for chain-coherent sampling (#2756)
tektoncd/pipelines-as-code: docs/content/docs/operations/tracing.md

Reviewer: Since the configmap based tracing would no longer work, can we add a callout for any existing users.
Something like [this](https://docs.pipelinesascode.com/v0.48.0/docs/operations/profiling/#enabling-profiling).

<img width="712" height="118" alt="Image" src="https://github.com/user-attachments/assets/b9071749-fcb0-47de-a581-79f9557e0658" />

Reviewer 2: I don't agree with "The old" starting because it's a living guide doc how to configure tracing in PaC so old and new thing should be mentioned in release notes not here and also it's not deprecation notice. something like this here would be good:

```
{{< hint warning >}}
PaC doesn't consider `tracing-protocol`, `tracing-endpoint`, and `tracing-sampling-rate` ConfigMap keys for tracing configuration. Configure tracing via OpenTelemetry environment variables (see below).
{{< /hint >}}
```

#### Sampler choice and chain coherency

The `parentbased_*` sampler family honors the parent span's sample decision carried in the W3C `traceparent` flag bit. When every service in the delivery chain uses parent-based samplers, the root span's sampling decision propagates end to end: each service either keeps its spans or drops them based on what the root chose. Flat-rate samplers (`traceidratio` without parent-based) cause each service to roll independently, which at fractional sampling fragments the chain into orphaned spans whose `parent_spanID` references a span that was dropped. `parentbased_always_on` keeps everything; `parentbased_traceidratio` with a numeric argument samples a coherent fraction.

### Via Knative observability ConfigMap

Set in `pipelines-as-code-config-observability`:

* `tracing-protocol` - `grpc`, `http/protobuf`, `stdout`, or `none`. Default: `none`.
* `tracing-endpoint` - Collector endpoint for `grpc` or `http/protobuf`.
* `tracing-sampling-rate` - Sample fraction. Per-component independent.

> ConfigMap changes are picked up live via Knative's configmap watcher.

Reviewer: This conflicts with the original behaviour documented
```
Changes to `tracing-protocol`, `tracing-endpoint`, and `tracing-sampling-rate` require restarting the controller and watcher pods. The trace exporter is created once at startup from the ConfigMap values at that time. Set `tracing-protocol` to `none` or remove the tracing keys to disable tracing.
```
Is that intentional?

### When both are configured

> OpenTelemetry takes precedence: all spans flow through the OpenTelemetry exporter. The Knative tracer is initialized at startup but receives no spans, leaving its exporter connection open for the process lifetime.

Reviewer: [nit]
> leaving its exporter connection open for the process lifetime.

This sounds a bit ominous to me, maybe we can tone it down `this is a minor resource overhead with no functional impact. `

## Enabling tracing

Pipelines-as-Code reads tracing configuration from standard OpenTelemetry environment variables on the controller and watcher pods:

* `OTEL_EXPORTER_OTLP_ENDPOINT`: OTLP collector endpoint URL. Required to enable tracing.
* `OTEL_TRACES_SAMPLER`: Sampler family. Required to enable tracing. Supported values: `always_on`, `always_off`, `traceidratio`, `parentbased_always_on`, `parentbased_always_off`, `parentbased_traceidratio`.
* `OTEL_TRACES_SAMPLER_ARG`: Numeric argument for ratio-based samplers. Set to `0.1` with `OTEL_TRACES_SAMPLER=parentbased_traceidratio` to sample 10% of root traces while keeping the chain coherent on those that are kept.
* `OTEL_EXPORTER_OTLP_PROTOCOL` (or traces-specific `OTEL_EXPORTER_OTLP_TRACES_PROTOCOL`): OTLP transport. Supported values: `grpc`, `http/protobuf`. Default: `grpc`.

Both `OTEL_EXPORTER_OTLP_ENDPOINT` and `OTEL_TRACES_SAMPLER` must be set to opt in to tracing. If either is unset PaC falls back to a noop tracer that emits no spans and incurs no exporter cost. Changes to any of these env vars take effect on the next pod restart.

> PaC honors inbound `traceparent` headers on incoming webhook requests via the W3C TraceContext propagator. OTel Baggage is intentionally not honored; per [ADR 0061](https://github.com/konflux-ci/architecture/blob/main/ADR/0061-distributed-tracing.md), this delivery-tracing chain does not use Baggage for cross-service attribute propagation because all required attributes are locally available at every emission point.

Reviewer: We should not link a downstream project's design doc in PaC's documentation.

Reviewer 2: is github.com/konflux-ci considered downstream?? I think NO it's another open source repository.

Reviewer: Downstream would be anything that builds on or utilizes PaC and if a change here is affecting konflux-ci, then it does qualify as a downstream project.
Even so, point still stands. We should not link a design doc from another project.

Reviewer 3: yes please remove konflux references, we are not a dep on this

> PaC honors inbound `traceparent` headers on incoming webhook requests via the W3C TraceContext propagator. OTel Baggage is intentionally not honored; per [ADR 0061](https://github.com/konflux-ci/architecture/blob/main/ADR/0061-distributed-tracing.md), this delivery-tracing chain does not use Baggage for cross-service attribute propagation because all required attributes are locally available at every emission point.

Reviewer: Broken link, get a 404

Reviewer 2: actually it's 0062-distributed-tracing.md

### Sampler choice and chain coherency

> The `parentbased_*` sampler family honors the parent span's sample decision carried in the W3C `traceparent` flag bit. When PaC, Tekton, integration-service, and release-service all use parent-based samplers, the root span's sampling decision propagates through the whole delivery chain: every service either keeps its spans or drops them based on what the root chose. Flat-rate samplers (`traceidratio` without parent-based) cause each service to roll independently, which at fractional sampling fragments the chain into orphaned spans whose `parent_spanID` references a span that was dropped. `parentbased_always_on` keeps everything; `parentbased_traceidratio` with a numeric argument samples a coherent fraction.

Reviewer: **[warning/docs]** Remaining Konflux-specific service references

The earlier reviewer feedback asked to remove Konflux references. The ADR link was removed, but "integration-service" and "release-service" on this line are Konflux-CI-specific service names that leaked through. PaC is upstream and shouldn't assume a specific deployment topology in its docs.

Consider generalizing:

```suggestion
The `parentbased_*` sampler family honors the parent span's sample decision carried in the W3C `traceparent` flag bit. When every service in the delivery chain uses parent-based samplers, the root span's sampling decision propagates end to end: each service either keeps its spans or drops them based on what the root chose. Flat-rate samplers (`traceidratio` without parent-based) cause each service to roll independently, which at fractional sampling fragments the chain into orphaned spans whose `parent_spanID` references a span that was dropped. `parentbased_always_on` keeps everything; `parentbased_traceidratio` with a numeric argument samples a coherent fraction.
```
